### DISCUSSION TITLE
The trait `Send` is not implemented: how to use boa in async threaded context best?

### DISCUSSION THREAD

**Discussion by k-bx** [boa-dev/boa#3093] [Q&A] [answered]
I've made a imple [repository showing the error](https://github.com/k-bx/play-boa-engine-send), the essence is very simple use case:

```rust
use boa_engine::{Context, Source};

#[tokio::main]
async fn main() {
    let j1 = tokio::spawn(worker());
    tokio::try_join!(j1).unwrap();
}

async fn worker() {
    let context = make_context();
    let boa_context = tokio::sync::RwLock::new(context);
    some_worker_func(&boa_context).await;
}

pub fn make_context<'context>() -> Context<'context> {
    let mut context = Context::default();
    let src2 = Source::from_bytes(r#"var two = 2;"#);
    let _ = context.eval(src2).unwrap();
    context
}

pub async fn some_worker_func(
    boa_context: &tokio::sync::RwLock<boa_engine::Context<'_>>,
) -> () {
    println!("do nothing");
}
```

This fails with many erros like this:

```
the trait `Send` is not implemented for `Rc<(dyn HostHooks + 'static)>`
...
the trait `Send` is not implemented for `Rc<(dyn JobQueue + 'static)>`
...
the trait `Send` is not implemented for `Rc<(dyn ModuleLoader + 'static)>`
...
the trait `std::marker::Sync` is not implemented for `(dyn HostHooks + 'static)`
...
```

Is this by design? What's the best way to go around this?

**Comment by jedel1043** [comment, marked as answer]
Short answer: Please use `tokio::task::spawn_local` instead.

Long answer: The JS execution model pretty much restricts all implementors of the spec to be single-threaded (with respect to the interpreter, but there are other ways to share work between threads like Workers), which is why Boa extensively uses `!Send` types and thread locals to maximize performance.

A Boa context is local to the current thread, so if you want to wrap it in a tokio task, you'll need to use the capabilities of tokio to run `!Send` futures: `task::spawn` for simple tasks and `LocalSet` for task groups.

**Comment by k-bx** [comment]
@jedel1043 thank you very much for the detailed explanation. Separate "thank you" for a timely one, since I've started reworking the whole thing to use a `bmrng`-channel around my boa engine calls which are not needed now :)
